[Wang Zhenlin] iP

Criss-Wang opened the pull request.

Commit:
Add Gradle support

Criss-Wang committed:
Add checkstyle.xml

Criss-Wang committed:
Update Duke.java main() method

Complete Level-1 Task.

Let's simulate the chat bot called "Jarvis" by simple while loop
which outputs the command from user and takes in a new one
until "bye" command is received.

Criss-Wang committed:
Level-2 Intermediate work

Criss-Wang committed:
Add a set of command-related and task classes

Make the command reading and processing systematic.

Let's,
* create a few command classes to separate different types of command
* create a CommandReader class to help with the separation process
* create a immutable Task class to denote the user's tasks
* create a immutable TaskList class to denote the list of tasks
* create a CommandAgent class to handle the command request
and generate feedback

Criss-Wang committed:
Change handleCommand() method

There will be more types of request, in such cases, if-else method is
less readable as compared to switch method.

Change the conditioning method to switch method makes the code cleaner.

As a result, let's condition on command's request and set actions and
feedback for each case respectively.

Criss-Wang committed:
Change constructors and methods for command class and its subclasses

Some content in the command class and its subclasses appear to be
redundant. The getContent() method can be unified.

In light of this, let's
* Change the plain userInput attribute of Command object to a more
meaningful content attribute
* Modify the constructor of Command class to initialize its content to
empty string and let ListCommand inherit this constructor
* Modify the constructor of other subclasses to initialize their content
to the proper content which excludes command keywords

Criss-Wang committed:
Change the parameter in switch method

Some user input have more than 1 word, with first word as keyword and
the rest being the content assoicated with the command

In light of this, let's
* split the user inputs into an array of string
* mark the first word as commandWord
* identify the correct type of command based on commandWord
* return the resultant command and its associate content

Criss-Wang committed:
Add attribute isDone in Task class

A status for the task is needed for the bot's requirement. Operations
on the status involves marking the task as done.

In light of this, let's:
* create the new attribute isDone
* add isDoneTask() and complete() methods to update the task's status
and display the status
* modify the toString() method to include the status information into
the string displayed

Criss-Wang committed:
Enable update of task status and extraction of task by id in TaskList

We need to the get task from the TaskList object and update it in the
list.

In light of this, let's:
* Add a getTaskById() method to extract the task via its id
* Add a markAsDone() method to update the task status in the list
* Add a generateFeedbackById() method to send a feedback to inform the
user that task has been done

Criss-Wang committed:
Add DoneCommand class

Handle the "done x" command by creating a new class for this type of
command, send the request as "update" for agent to handle x-th task.

Criss-Wang committed:
Complete Level-3

Criss-Wang committed:
Add cases in CommandReader, generate more detailed content

There are more types of command, and content each input contains becomes
more versatile.

In light of this, let's:
* add case for each new command "todo", "deadline" and "event"
respectively
* extract the non-commandWord content from user input via a function
generateContent()

Criss-Wang committed:
Add notation for generateContent() function

Criss-Wang committed:
Handle the 3 subtypes of Task class

There are many classes related to tasks. There are more attributes
in the 3 subclasses, making their constructor and toString() method to
vary from original Task class. The feedback generation in TaskList looks
more reasonable if it is done in commandAgent.

Let's:
* create a package task to contain all the Task-related classes
* modify the constructor and toString() methods in Deadline, Event, Todo
classes
* add a getSize() method in TaskList to calculate size of the list
* remove the generateFeedbackById() method in TaskList

Criss-Wang committed:
Modify code to comply with code style

Criss-Wang committed:
Clear up DukeBuffer

Criss-Wang committed:
Create new command subclasses

The event/todo/deadline commands are particularly the subclass of
AddCommand which creates a task. Their content will have the task name,
some even have the schedule. Different content extraction and handling
method are needed.

In light of this, let's:
* create 3 subclasses of AddCommand
* modify the getContent() method in each command class so that
respective content are stored in a list of strings

Criss-Wang committed:
Add feedback generation methods and restructure handleCommand() code

The feedback displayed can be decorated to look better. Feedback
generation should be done in this class. The creation of a task can also
be done in the agent as command should not have task creation ability.

In light of this, let's create feedback generation methods for each of
the creation/update/retrieval. Make sure the taskList is updated within
the handleCommand() method and respective feedback are produced for each
command request.

Criss-Wang committed:
Deadline: the identifier in toString() should be D not E

Let's fix the bug by changing the label to D

Criss-Wang committed:
Edit test code and fix bug in Deadline class after test

Another issue detected after test shows "at" is used in Deadline class.

The solution is to just change the "at" to "by".

Criss-Wang committed:
Adjust code to corret style and add DukeException

Some parts of the code still have minor styling issue.
Creat an Exception class DukeException to represents some
command-related issues. But all the command classes are in the command
package. Cannot directly access outer package method. Also the
DukeException needs to be handled at the right place.

To fix the issues, let's:
* create a package for exception classes, called "exception"
* handle the DukeException in commandRead class, which produces a
ErrorCommand object upon catching the exception
* handle the ErrorCommand by outputing its error message as the content
in the command

Criss-Wang committed:
Identify various errors in command

The todo/deadline/event command produces error if command is missing
task name or schedule or the delimiter like /by and /at are not
detected.

To make these error handling work, let's detect the error in
CommandReader when we construct the command.

Criss-Wang committed:
Wrap all the code under a duke package

Have a package wrapping around the entire source code helps class from
inner packages to access the methods/attribute of classes from outer
packages easily.

In light of this, let's create a duke folder to contain the command and
task packages, as well as Duke, DukeBuffer, DukeException, CommmandAgent
and CommandReader class.

Criss-Wang committed:
Modify comments in all .java files

Upon refactoring of code into the duke package, some comments are
automatically modified by intellij, which is undesriable.

Let's undo these modification by deleting the "duke." prefix in the
comment sections of each method/class.

Criss-Wang committed:
Handle some left-over modification and create delete command handling

There are some comments whose "duke." prefix are not removed.
Work on level-6, create the scenario for "delete" command.

Let's clean up the code first, and create a DeleteCommand with Error
Handling behaviour. Add a case in CommandReader to parse the delete
input as well.

Criss-Wang committed:
Handle delete task logic and feedback generation

The task needs to be deleted should be shown in feedback. The tasklist
needs to send the task to CommandAgent before it gets deleted.

Let's create a case for deletion of the task by its index, keep a copy
of the task before it gets deleted in the taskList. Use the copy and
the updated List for feedback generation.

Criss-Wang committed:
Resolve sad face symbol's imcompatability with GBK encoding

The sad face in the error handling messge cause compile issue as the
terminal cannot resolve the symbol. There are some leftover "duke."
prefix in EXPECTED.TXT which causes test failure.

Let's remove the sad faces in the error message for now and clean up
the EXPECTED.TXT, and enforce file encoding in runtest.bat to UTF-8.

Criss-Wang committed:
Change isExit() method in Command to static method in ExitCommand

The isExit() method examines the content of the command, which will
does not use a getter. I consider the check of Exit command more natural
when it is a method always associated with the ExitCommand Class.

Hence, let's remove the isExit() command in Command class and replace
its functionality with a static method isExitCommand() in ExitCommand
class. Replace the exit condition in do-while loop in Duke class with
the isExitCommand check as well.

Criss-Wang committed:
Modify the Task class to an abstract class

The Task class is separated into 3 different types, a user input should
always specify one of the 3 types. Hence it is unlikely to create a new
Task object directly.

In light of this, let's make the Task class an abstract class, change
the complete() method in Task class to abstract method as well.

Criss-Wang committed:
Add formating function for Task objects and TaskList

The task representation in the data/duke.txt file has a specific format.
Additional functions are needed to convert each task in the task list
into that format.

In light of this, let's:
* create a format() method for Task objects and implement them in each
of the Deadline/Todo/Event subclasses based on their attributes.
* create a formatTaskList() method for the taskList to format all the
tasks in the list.

Criss-Wang committed:
Create a Storage class to manage the task data I/O

The Storage class needs to handle save and load operations on data.
When the file/folder to store the task data are not present, the class
also need to detect them and create the correct file/folder immediately
upon loading of data file.

In light of this, let's:
* build a constructor for Storage class to store the file's folder path
and file path in relative path format
* create a createPath() method to create new path for data storage
if current path specified is not detected
* create save() and load() methods for data file reading/writing

Criss-Wang committed:
Add Storage object to relevant classes

We need to initialize the storage in Duke, and let the CommandAgent
to handle the I/O of task data via Storage object.

In light of this, let's:
* Add an instance of Storage into Duke, pass the instance into a
commantAgent
* Add the data I/O operation after each command in the agent via the
Storage class methods

Criss-Wang committed:
Merge branch 'branch-Level-7'

Criss-Wang committed:
Convert schedule data type from String to LocalDate

The schedule data needs to be converted from plain string to LocalDate
data, respective parser and formatter to output the LocalDate as String
are needed.

In light of this, let's:
* modify the constructors in each task subclasses to take in LocalDate
data
* Allow the LocalDate data to be formatted into string in these classes
* In Storage class, parse the tasks schedule in duke.txt file from
String type to LocalDate type

Criss-Wang committed:
Merge branch 'branch-Level-8'

Criss-Wang committed:
Removal of unused files and renaming of CommandReader.java

After reading the project requirement, it becomes clear that DukeBuffer
are now simply the TaskList, hence it provides little utility. The
ErrorCommand class can be removed as the error handling in the code
becomes more fluent. The CommandReader class should be renamed as Parser
as the level-9 requirement suggests.

In light of this, let's remove the 2 files and rename the CommandReader
class into Parser class, with its parse() method turned into static to
make Parser a utility class.

Criss-Wang committed:
Re-organize the code logic to follow the standard in level-9

The main logic in Duke.java file provides a sample reference for the
logic of the chat bot. We can follow this logic standard and adjust
our main logic correspondingly.

In light of this, let's:
* create a new class, Ui, for message output
* modify the Storage constructor to allow it to take in the data file
path
* tweak the Parser class to better handle exceptions when reading
command
* retrieve the old isExit() checking method via command content
* rename the feedback into response for the common request-response
styling
* modify the Duke.java class to fit the sample code in level-9
instruction

Criss-Wang committed:
Current packages maintenance and JavaDoc supplementation

The current packages are rather complete. Some minor issues like the
Event and Deadline constructor with schedule parameter in the string
representation can be removed as LocalDate type can be created when
the task is created. Some public method's JavaDoc can also be added.

Let's drop the constructor mentioned above, and add in a few more
JavaDocs for public methods.

Criss-Wang committed:
Setup JUnit test and created relevant directories

Criss-Wang committed:
Add test cases for all Task subclasses methods

The 3 classes have relatively straight-forward methods to test on.
More test on complicated classes like Command classes and Storage
will come up later.

Criss-Wang committed:
Create Jar file for project

Criss-Wang committed:
Adjust printTask() string output

The response header should be part of the response generated from
CommandAgent. TaskList should only output a list of task in printTask
method.

Criss-Wang committed:
Add Find command logic

Only allow 1 word as keyword per finding. Test has some bug to be fixed.

Criss-Wang committed:
Adjust automated testing scripts

Criss-Wang committed:
Fix error message parsing issue

Found the root cause for error parsing the unhappy face in error
messages. Add in more exception handling after find() function is
created, some tasks are already entered, some tasks are already done,
etc...

In light of this, let's:
* get the unhappy face back into the exception messages
* add exception handling for entering recorded tasks and complete done
tasks

Criss-Wang committed:
Merge branch 'branch-Level-9' and resolve merge conflicts

Criss-Wang committed:
Edit JavaDoc in the project

Modify the comments in JavaDoc to make them look logical.

Criss-Wang committed:
Merge branch 'branch-A-JavaDoc'

Criss-Wang committed:
Modify JavaDoc comments to stick to coding style

Criss-Wang committed:
Examine coding style and follow the coding standard

Criss-Wang committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Storage.java
resolved.

Criss-Wang committed:
Edit JavaDoc in some methods

Some methods' JavaDoc are not clear enough. A few methods related
to find command do not have JavaDoc.

Hence, let's add the JavaDoc for those missing, and make the annotation
clearer for these methods' JavaDoc.

Criss-Wang committed:
Update final attribute naming

The current naming of keyword length does not comply with coding
standard.

We need to Capitalize all the alphabets.

Let's update the naming of the keyword length attribute in FindCommand.

Criss-Wang committed:
Merge branch 'add-gradle-support' of https://github.com/Criss-Wang/ip into A-Gradle

Criss-Wang committed:
Build with Gradle

Criss-Wang committed:
Merge branch 'A-Gradle'

Criss-Wang committed:
Add JavaDoc for Storage and Task classes

Some classes' public constructors do not have JavaDoc.

The constructors for these classes need more interpretability via
documentation.

Let's add the JavaDoc for these constructors explaining the parameters.

Criss-Wang committed:
Merge branch 'A-Checkstyle'

# Conflicts:
#	gradle/wrapper/gradle-wrapper.jar
#	gradle/wrapper/gradle-wrapper.properties
#	gradlew
#	gradlew.bat

Criss-Wang committed:
Create first draft of JavaFX for level-10

Criss-Wang committed:
Merge branch 'branch-Level-10'

# Conflicts:
#	build.gradle
#	data/duke.txt
#	gradle/wrapper/gradle-wrapper.jar
#	gradle/wrapper/gradle-wrapper.properties
#	gradlew
#	gradlew.bat

Criss-Wang committed:
Adjust padding in dialog box

The dialog box has its text too close to scroll bar when the length
of the text exceeds one line.

Increase the right padding for the interface will solve the issue.

Let's increase the right padding from 5 to 20 and left padding from 5
to 10 for dialog box coherence.

Criss-Wang committed:
Review code and rectify coding quality violations

Some parts of current code do not comply with coding quality constraint
such as "Magic Number" avoidance, naming and nested conditioning.

These violations should be corrected to make the overall code more
robust and informative.

In light of this, let's:
* rename those inappropriate methods and variables
* bring the conditioning in if-else clause out, thus creating 2
conditioning clauses on the same level
* change request keyword into static variables attached to each command
type
* extract datetime formatter out as a standard pattern for task classes
* rename the format() method to avoid confusion with the Datetime's
in-built format method.

bangyiwu commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,69 @@
+     * @throws DukeException If the command construction involves error or
+     * the DateTime parsing involves error.
+     */
+    public static Command parse(String userInput) throws DukeException {
```
maybe a better function name here would be better, something like parseCommand

bangyiwu commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,41 @@
+     * @param content The task information supplied by the user.
+     * @throws DukeException If the content has missing name or schedule or the format does not contain /by.
+     */
+    public DeadlineCommand(String content) throws DukeException {
```
if possible, best to avoid arrowhead coding here

Criss-Wang replied:
Thanks for the`sharp observation. Just rectified the issue!

bangyiwu reviewed (commented):
Minor nitpicking on coding quality

chunyongg commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,121 @@
+        List<Task> tasksLoaded = new ArrayList<>();
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM d yyyy", Locale.ENGLISH);
+
+        for (String s : tasksInFile) {
```
Perhaps it would be great if there are some comments on how the parsing is done

chunyongg commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,70 @@
+
+    /**
+     * Returns a welcome message in the chat bot.
+     * @return The greeting String.
+     */
+    public static String greet() {
+        return "Hello! I'm Jarvis\n"
+                + "What can I do for you?";
+    }
+
+    /**
+     * Prints welcome message in the chat bot.
+     */
+    public void showWelcome() {
+        System.out.println("Hello! I'm Jarvis.\n"
+                + "What can I do for you?");
+    }
+
```
For this segment, consider using the return String from greet() and printing it in showWelcome()

Criss-Wang replied:
Thanks for the suggestion!

chunyongg commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,41 @@
+        if (!content.contains("/by")) {
+            throw new DukeException("☹ OOPS!!! The information contains invalid delimiter");
+        } else {
+            String[] contentParts = content.split(" /by ");
+            if (contentParts[0].equals("") || contentParts[1].equals("")) {
```
Perhaps giving a comment on the expected format of the input (so that contentParts[0] and contentParts[1] is understandable to reviewers)

chunyongg reviewed (commented):
Overall, the code is nicely encapsulated into small units and understandable. Method names, variable names, boolean names etc also give a good idea as to what they stand for.

Criss-Wang committed:
Resolve code quality issue from Pull Request feedback

Criss-Wang committed:
Customization for GUI

Criss-Wang committed:
Add in assertions

Some parts of the code implicitly assume the length of an array/list
to be greater than a certion number.

The assumption needs to be verified beforehand to prevent
indexOutOfBound issue.

Let's implement assertions on the size of these list/array.

Criss-Wang committed:
Examine code quality issue, refactor code

There are several parts of code quality violations:
* SLAP is not followed in Duke.java
* executeCommand() method in CommandAgent.java is too long
* Some naming issues

Let's rectify the naming issues, split the loop out from run() method
into runUnitlExit(), and create methods specifically for each command
execution to shorten the executeCommand() method.

Criss-Wang committed:
Merge pull request #1 from Criss-Wang/branch-A-Assertions

Add in assertions

Criss-Wang committed:
Merge branch 'master' into branch-A-CodeQuality

Criss-Wang committed:
Merge pull request #2 from Criss-Wang/branch-A-CodeQuality

Examine code quality issue, refactor code

Criss-Wang committed:
Examine current lambda expression, verify stream expression

Criss-Wang committed:
Verify current stream working properly

Criss-Wang committed:
Merge pull request #3 from Criss-Wang/branch-A-Lambdas

Examine current lambda expression, verify stream expression

Criss-Wang committed:
Merge branch 'master' into branch-A-Streams

# Conflicts:
#	src/main/java/duke/task/TaskList.java

Criss-Wang committed:
Merge pull request #4 from Criss-Wang/branch-A-Streams

Verify current stream working properly

Criss-Wang committed:
Create gradle.yml

Criss-Wang committed:
Rectify minor styling issue

Criss-Wang committed:
Merge branch 'branch-A-CI' of https://github.com/Criss-Wang/ip into branch-A-CI

Criss-Wang committed:
Merge branch 'branch-A-CI'

Criss-Wang committed:
Add new functionality: reminder command

The command asks for user input in the form of "remind + number" where
the number is the number of tasks not done and whose schedule are most
recent. The tasks displayed are sorted in their schedule order, with
the tasks whose schedule has past the day getting indicated with an
expiry sign.

Criss-Wang committed:
Merge branch 'branch-B-Reminders'

Criss-Wang committed:
Test Github Action issue

Criss-Wang committed:
Add inheritDoc for overriding method and lambda expression

Some public overriding methods can have javadoc inheriting from the
overriden method. More lambda expression can be applied on List objects.

Let's use {@inheritCode} clause in JavaDoc for those overriding methods
and add lambda expression in list object using .forEach() method.

Criss-Wang committed:
Verify GUI, create Jar file and do smoke test

Criss-Wang committed:
Fix minor bug

Criss-Wang committed:
Update User Guide and fix GUI issue

The application cannot be opended with .jar file by double clicking on
it. User guide can have more fancy design.

The issue can be solved by modifying the build.gradle and manifest's
main class to 'duke.Launcher'.

Let's add more styles to the User Guide and solve the GUI issue
accordingly.

Criss-Wang committed:
Further update README.md

Criss-Wang committed:
Further update README.md

Criss-Wang committed:
Edit styling in User Guide

Criss-Wang committed:
Fix image issue

Criss-Wang committed:
Fix spacing issue in UserGuide

Criss-Wang committed:
Update README.md

Criss-Wang committed:
Update README.md

Criss-Wang committed:
Finalize User Guide

Criss-Wang committed:
Upload Ui.png and do smoke testing

Criss-Wang committed:
More smoke testing and bug fixing
